File: frontend/src/api/report.ts
import client from './client'

// --- Report Types ---

export interface OutlookData {
  timeframe: string
  bull_case: string
  bull_target: number
  bull_probability: number
  base_case: string
  base_target: number
  base_probability: number
  bear_case: string
  bear_target: number
  bear_probability: number
}

export interface PriceLevel {
  price: number
  source: string
  strength: number
}

export interface FibonacciData {
  swing_high: number
  swing_low: number
  direction: string
  retracement: Record<string, number>
  extension: Record<string, number>
}

export interface PriceTarget {
  scenario: string
  target_price: number
  upside_pct: number
  rationale: string
  timeframe: string
  confidence: number
}

export interface ReportResult {
  stock_code: string
  stock_name: string
  report_date: string
  current_price: number
  // Price performance
  price_change_1w: number | null
  price_change_1m: number | null
  price_change_3m: number | null
  price_change_6m: number | null
  price_change_1y: number | null
  high_52w: number | null
  low_52w: number | null
  pct_from_52w_high: number | null
  pct_from_52w_low: number | null
  // Trend
  trend_direction: string
  trend_strength: string
  momentum_status: string
  volatility_level: string
  overall_rating: string
  ma_alignment: string
  // Support / Resistance
  support_levels: PriceLevel[]
  resistance_levels: PriceLevel[]
  fibonacci: FibonacciData | null
  price_targets: PriceTarget[]
  // Momentum indicators
  adx_value: number | null
  adx_interpretation: string
  rsi_value: number | null
  rsi_interpretation: string
  macd_value: number | null
  macd_signal_value: number | null
  macd_histogram: number | null
  macd_interpretation: string
  k_value: number | null
  d_value: number | null
  kd_interpretation: string
  // Volume
  volume_trend: string
  volume_ratio: number | null
  volume_interpretation: string
  // Volatility
  atr_value: number | null
  atr_pct: number | null
  bollinger_width: number | null
  bollinger_position: number | null
  volatility_interpretation: string
  // Risk
  max_drawdown_1y: number | null
  current_drawdown: number | null
  risk_reward_ratio: number | null
  risk_interpretation: string
  // Outlook
  outlook_3m: OutlookData | null
  outlook_6m: OutlookData | null
  outlook_1y: OutlookData | null
  // Summary
  summary_text: string
  // Strategy
  v4_analysis: Record<string, unknown> | null
  v2_analysis: Record<string, unknown> | null
  // Fundamentals
  fundamentals: Record<string, unknown> | null
  fundamental_interpretation: string | null
  fundamental_score: number | null
  analyst_data: Record<string, unknown> | null
  // News
  news_items: Record<string, unknown>[] | null
  news_sentiment_score: number | null
  news_sentiment_label: string | null
  news_insights: string | null
  news_themes: string[] | null
  // Actionable
  actionable_recommendation: string | null
  industry_risks: string[] | null
  technical_conflicts: string[] | null
  technical_bias: string | null
  peer_context: Record<string, unknown> | null
  valuation: Record<string, unknown> | null
  // Institutional
  institutional_score: number | null
  is_biotech: boolean
  rating_weights: Record<string, number> | null
  cash_runway: Record<string, unknown> | null
  rating_decision: Record<string, unknown> | null
}

export const reportApi = {
  generate: (code: string, periodDays = 730, marketRegime?: string) =>
    client.post<any, ReportResult>(`/report/${code}/generate`, {
      period_days: periodDays,
      market_regime: marketRegime,
    }),
}

File: backend/routers/report.py
"""分析報告路由"""

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

router = APIRouter()


class ReportRequest(BaseModel):
    period_days: int = 730
    market_regime: str | None = None


@router.post("/{code}/generate")
def generate_report_api(code: str, req: ReportRequest):
    """產生完整分析報告"""
    from analysis.report import generate_report
    from backend.dependencies import make_serializable
    try:
        report = generate_report(
            code,
            period_days=req.period_days,
            market_regime=req.market_regime,
        )

        # 序列化 ReportResult dataclass
        result = {
            "stock_code": report.stock_code,
            "stock_name": report.stock_name,
            "report_date": report.report_date.isoformat(),
            "current_price": report.current_price,
            # 價格表現
            "price_change_1w": report.price_change_1w,
            "price_change_1m": report.price_change_1m,
            "price_change_3m": report.price_change_3m,
            "price_change_6m": report.price_change_6m,
            "price_change_1y": report.price_change_1y,
            "high_52w": report.high_52w,
            "low_52w": report.low_52w,
            "pct_from_52w_high": report.pct_from_52w_high,
            "pct_from_52w_low": report.pct_from_52w_low,
            # 趨勢
            "trend_direction": report.trend_direction,
            "trend_strength": report.trend_strength,
            "momentum_status": report.momentum_status,
            "volatility_level": report.volatility_level,
            "overall_rating": report.overall_rating,
            "ma_alignment": report.ma_alignment,
            # 支撐壓力
            "support_levels": [
                {"price": s.price, "source": s.source, "strength": s.strength}
                for s in (report.support_levels or [])
            ],
            "resistance_levels": [
                {"price": r.price, "source": r.source, "strength": r.strength}
                for r in (report.resistance_levels or [])
            ],
            # 費氏
            "fibonacci": {
                "swing_high": report.fibonacci.swing_high,
                "swing_low": report.fibonacci.swing_low,
                "direction": report.fibonacci.direction,
                "retracement": report.fibonacci.retracement,
                "extension": report.fibonacci.extension,
            } if report.fibonacci else None,
            # 目標價
            "price_targets": [
                {
                    "scenario": t.scenario,
                    "target_price": t.target_price,
                    "upside_pct": t.upside_pct,
                    "rationale": t.rationale,
                    "timeframe": t.timeframe,
                    "confidence": t.confidence,
                }
                for t in (report.price_targets or [])
            ],
            # 動能指標
            "adx_value": report.adx_value,
            "adx_interpretation": report.adx_interpretation,
            "rsi_value": report.rsi_value,
            "rsi_interpretation": report.rsi_interpretation,
            "macd_value": report.macd_value,
            "macd_signal_value": report.macd_signal_value,
            "macd_histogram": report.macd_histogram,
            "macd_interpretation": report.macd_interpretation,
            "k_value": report.k_value,
            "d_value": report.d_value,
            "kd_interpretation": report.kd_interpretation,
            # 量能
            "volume_trend": report.volume_trend,
            "volume_ratio": report.volume_ratio,
            "volume_interpretation": report.volume_interpretation,
            # 波動
            "atr_value": report.atr_value,
            "atr_pct": report.atr_pct,
            "bollinger_width": report.bollinger_width,
            "bollinger_position": report.bollinger_position,
            "volatility_interpretation": report.volatility_interpretation,
            # 風險
            "max_drawdown_1y": report.max_drawdown_1y,
            "current_drawdown": report.current_drawdown,
            "risk_reward_ratio": report.risk_reward_ratio,
            "risk_interpretation": report.risk_interpretation,
            # 展望
            "outlook_3m": _serialize_outlook(report.outlook_3m),
            "outlook_6m": _serialize_outlook(report.outlook_6m),
            "outlook_1y": _serialize_outlook(report.outlook_1y),
            # 摘要
            "summary_text": report.summary_text,
            # 策略
            "v4_analysis": report.v4_analysis,
            "v2_analysis": report.v2_analysis,
            # 基本面
            "fundamentals": report.fundamentals,
            "fundamental_interpretation": report.fundamental_interpretation,
            "fundamental_score": report.fundamental_score,
            "analyst_data": report.analyst_data,
            # 消息面
            "news_items": report.news_items,
            "news_sentiment_score": report.news_sentiment_score,
            "news_sentiment_label": report.news_sentiment_label,
            "news_insights": report.news_insights,
            "news_themes": report.news_themes,
            # 行動建議
            "actionable_recommendation": report.actionable_recommendation,
            "industry_risks": report.industry_risks,
            "technical_conflicts": report.technical_conflicts,
            "technical_bias": report.technical_bias,
            "peer_context": report.peer_context,
            "valuation": report.valuation,
            # 籌碼面（Gemini R19）
            "institutional_score": report.institutional_score,
            "is_biotech": report.is_biotech,
            "rating_weights": report.rating_weights,
            # Cash Runway（Gemini R20）
            "cash_runway": report.cash_runway,
        }

        return make_serializable(result)
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


def _serialize_outlook(outlook) -> dict | None:
    if outlook is None:
        return None
    return {
        "timeframe": outlook.timeframe,
        "bull_case": outlook.bull_case,
        "bull_target": outlook.bull_target,
        "bull_probability": outlook.bull_probability,
        "base_case": outlook.base_case,
        "base_target": outlook.base_target,
        "base_probability": outlook.base_probability,
        "bear_case": outlook.bear_case,
        "bear_target": outlook.bear_target,
        "bear_probability": outlook.bear_probability,
    }
